### PR TITLE
ROE-2313 Trim leading zeros from some date fields

### PR DESCRIPTION
**Pull request by markpit**
### JIRA link
https://companieshouse.atlassian.net/browse/ROE-2313

### Change description
Trimming any leading zeros from date values to prevent errors when saving to the Api

### Work checklist

- [ ] Tests added where applicable
- [ ] UI changes meet accessibility criteria

### Merge instructions

We are committed to keeping commit history clean, consistent and linear. To achieve this, this commit should be structured as follows:

```
<type>[optional scope]: <description>
```

and contain the following structural elements:

- fix: a commit that patches a bug in your codebase (this correlates with PATCH in semantic versioning),
- feat: a commit that introduces a new feature to the codebase (this correlates with MINOR in semantic versioning),
- BREAKING CHANGE: a commit that has a footer `BREAKING CHANGE:` introduces a breaking API change (correlating with MAJOR in semantic versioning). A BREAKING CHANGE can be part of commits of any type,
- types other than `fix:` and `feat:` are allowed, for example `build:`, `chore:`, `ci:`, `docs:`, `style:`, `refactor:`, `perf:`, `test:`, and others,
- footers other than `BREAKING CHANGE: <description>` may be provided.

### CODE REVIEW

**Comment by markpit** [comment]
> As a general comment, I can't see any unit tests for the particular scenarios described in the JIRA ticket - valid days and months but with leading zeroes. Are these missing or due to come later?

I thought the 'missing x when x is zero' would cover that scenario but I can look at adding tests or improve existing ones to check the zero is in fact removed.

File: src/validation/fields/date.validation.ts [hunk 1]
```diff
@@ -25,6 +25,12 @@ import { conditionalDateValidations, dateContext, dateContextWithCondition, date
 // to prevent more than 1 error reported on the date fields we check if the year is valid before doing some checks.
 // This means that the year check is checked before some others
 export const start_date_validations = [
+  body("start_date-day")
+    .ltrim("0"),
```
[the remark this answers is not in the document]
Comment (markpit) [reply src/validation/fields/date.validation.ts:29]: i was thinking of adding comments to say that the custom checks actually look at all 3 parts of the dates so we need to trim the zeros up front before any custom validation.

**Comment by mwestacott** [comment]
> > As a general comment, I can't see any unit tests for the particular scenarios described in the JIRA ticket - valid days and months but with leading zeroes. Are these missing or due to come later?
> 
> I thought the 'missing x when x is zero' would cover that scenario but I can look at adding tests or improve existing ones to check the zero is in fact removed.

So yes, the 'when x is zero' part is not the scenario in the ticket, it's 'when x has leading zeroes'.

File: src/validation/fields/date.validation.ts [hunk 2]
```diff
@@ -25,6 +25,12 @@ import { conditionalDateValidations, dateContext, dateContextWithCondition, date
...
+  body("start_date-month")
+    .ltrim("0"),
```
[the remark this answers is not in the document]
Comment (markpit) [reply src/validation/fields/date.validation.ts:31]: you can type non numeric chars into the date field but when you submit the page, the web validation will stop you and ask for a real date to be entered.

File: src/validation/fields/date.validation.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (mwestacott) [reply src/validation/fields/date.validation.ts:29]: Aha, that's why! Then yes, a comment would be helpful.

[the remark this answers is not in the document]
Comment (markpit) [reply src/validation/fields/date.validation.ts:30]: comments updated

File: test/controllers/beneficial.owner.gov.controller.spec.ts [hunk 3]
```diff
@@ -460,13 +460,49 @@ describe("BENEFICIAL OWNER GOV controller", () => {
...
+      expect(resp.text).toContain(ErrorMessages.YEAR);
+      expect(resp.text).not.toContain(ErrorMessages.INVALID_DATE);
       expect(resp.text).not.toContain(ErrorMessages.DATE_NOT_IN_PAST_OR_TODAY);
       expect(mockSaveAndContinue).not.toHaveBeenCalled();
     });
 
+    test(`leading zeros are stripped from start date`, async () => {
+      mockPrepareData.mockReturnValueOnce(BENEFICIAL_OWNER_GOV_OBJECT_MOCK);
+      const beneficialOwnerGov = { ...BENEFICIAL_OWNER_GOV_BODY_OBJECT_MOCK_WITH_ADDRESS };
+      beneficialOwnerGov["start_date-day"] = "0030";
```
Comment (mwestacott) [inline test/controllers/beneficial.owner.gov.controller.spec.ts:492]: Do you need to add a test (for example) when leading zeroes feature in a non-valid day, month or year. E.g. on this line, what happens when `start_date-day` is "0032"?

File: test/controllers/beneficial.owner.gov.controller.spec.ts [hunk 4]
```diff
@@ -440,18 +440,18 @@ describe("BENEFICIAL OWNER GOV controller", () => {
...
       expect(resp.text).not.toContain(ErrorMessages.MONTH);
       expect(resp.text).not.toContain(ErrorMessages.YEAR);
-      expect(resp.text).toContain(ErrorMessages.INVALID_DATE);
+      expect(resp.text).not.toContain(ErrorMessages.INVALID_DATE);
       expect(resp.text).not.toContain(ErrorMessages.DATE_NOT_IN_PAST_OR_TODAY);
       expect(mockSaveAndContinue).not.toHaveBeenCalled();
     });
 
-    test(`renders the current page ${config.BENEFICIAL_OWNER_GOV_PAGE} with only INVALID_DATE error when month is zero`, async () => {
+    test(`renders the current page ${config.BENEFICIAL_OWNER_GOV_PAGE} with only missing MONTH error when month is zero`, async () => {
```
[the remark this answers is not in the document]
Comment (markpit) [reply test/controllers/beneficial.owner.gov.controller.spec.ts:451]: I'll update the ticket

File: test/controllers/beneficial.owner.gov.controller.spec.ts (hunk 3 above)
Comment (markpit) [reply test/controllers/beneficial.owner.gov.controller.spec.ts:492]: just tested this and eg start date of 0032 still returns 'invalid date' error even when the "0" stripping is not added. So we will still get the same error if entering 0032 whether it strips the zeroes or not. So i don't think adding tests for those scenarios will prove anything.

Comment (markpit) [reply test/controllers/beneficial.owner.gov.controller.spec.ts:492]: We can test the year length as 0040 will fail validation with "0" trimming (length = 2 when validated) and will pass validation without "0" trimming as length will be 4 when validated.

Comment (mwestacott) [reply test/controllers/beneficial.owner.gov.controller.spec.ts:492]: Sounds good to me. The only reason for adding tests in these scenarios was because the scenario wasn't tested at all and if we just assert on what it does now, we'll then pick up any potential breakages or regressions in the future (when the date logic changes again...).

Comment (markpit) [reply test/controllers/beneficial.owner.gov.controller.spec.ts:492]: ok i can add the extra tests , won't hurt

**Review by mwestacott — APPROVED** [review]
Looks good to me.

**Comment by dlloyd1-ch** [comment]
Could we use the 'maxlength' attribute in the input field to prevent the issue happening in the first place?

**Comment by markpit** [comment]
> Could we use the 'maxlength' attribute in the input field to prevent the issue happening in the first place?

that could be a better solution, I can try that out

**Comment by markpit** [comment]
> Could we use the 'maxlength' attribute in the input field to prevent the issue happening in the first place?

Looking into using maxlength, the govuk nunjucks date component doesn't seem to support it and further digging led me to this question on the alphagov github site, where they don't recommend maxlength, but instead suggest displaying an error to the user. This PR doesn't match this approach (it auto trims the leading zeros) so perhaps solution might need to be re-done to display message to user instead - https://github.com/alphagov/govuk-design-system/issues/1977
